src/AWSMetadataUtilities.jl

```diff
@@ -0,0 +1,578 @@
+    end
+
+    return uri
+end
+
+"""
+    _get_function_parameters(input, shapes)
+
+Get the required and optional parameters for a given operation.
+
+# Arguments
+- `input::String`: The input object for the given operation
+- `shapes::Dict{String, Any}`: Dictionary of all shapes for this AWS Service
+
+# Returns
+- `Tuple(Dict, Dict)`: (required_parameters, optional_parameters)
+"""
+function _get_function_parameters(input::String, shapes::Dict)
+    """
+        _get_parameter_name(parameter::String, input_shape::AbstractDict){String, <:Any}
+
+    Find the correct parameter name for making requests. Certain ones have a specific locationName, for Batch requests this locationName is nested one shape deeper.
+
+    # Arguments
+    - `parameter::String`: Name of the original parameter
+    - `input_shape::AbstractDict{String, <:Any}`: The parameter shape
+    
+    # Returns
+    - `String`: Either the original parameter name, the locationName for the parameter, or the locationName nested one shape deeper
+    """
+    function _get_parameter_name(parameter::String, input_shape::AbstractDict{String, <:Any})
+        # If the parameter has a locationName, return it
+        if haskey(input_shape["members"][parameter], "locationName")
+            return input_shape["members"][parameter]["locationName"]
+        end
+
+        # Check to see if the nested shape has a locationName
+        nested_shape = input_shape["members"][parameter]["shape"]
+        nested_member = get(shapes[nested_shape], "member", "")
+
+        # If nested_shape[member] exists return locationName (if exists), otherwise return the original parameter name
+        return !isempty(nested_member) ? get(nested_member, "locationName", parameter) : parameter
+    end
+
+    required_parameters = LittleDict{String, Any}()
+    optional_parameters = LittleDict{String, Any}()
+    input_shape = shapes[input]
+
+    if haskey(input_shape, "required")
+        for parameter in input_shape["required"]
+            parameter_name = _get_parameter_name(parameter, input_shape)
+
+            # Check if the parameter needs to be in a certain place
+            parameter_location = get(input_shape["members"][parameter], "location", "")
+
+            documentation = _clean_documentation(get(input_shape["members"][parameter], "documentation", ""))
+
+            required_parameters[parameter_name] = LittleDict{String, String}(
+                "location" => parameter_location,
+                "documentation" => documentation,
+            )
+        end
+    end
+
+    if haskey(input_shape, "members")
+        for parameter in input_shape["members"]
+            parameter_name = get(input_shape["members"][parameter[1]], "locationName", parameter[1])
+
+            if !haskey(required_parameters, parameter_name)
+                documentation = _clean_documentation(get(parameter[2], "documentation", ""))
+                idempotent = get(parameter[2], "idempotencyToken", false)
```

Comment:
Variable names could use some work beyond what I suggested:
```suggestion
        for (member_key, member_value) in input_shape["members"]
            parameter_name = get(input_shape["members"][member_key], "locationName", member_key)

            if !haskey(required_parameters, parameter_name)
                documentation = _clean_documentation(get(member_value, "documentation", ""))
                idempotent = get(member_value, "idempotencyToken", false)
```